### PR TITLE
[MXNET-978] Higher Order Gradient Support `sinh`, `cosh`.

### PR DESCRIPTION
**Pull request by kshitij12345**
## Description ##
PR intends to add support for higher order gradient for `sinh`, `cosh`.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA-978 issue](https://issues.apache.org/jira/browse/MXNET-978) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] higher order gradient for a `sinh`, `cosh`.
- [x] unit test for the same.

### CODE REVIEW

**Comment by anirudhacharya** [comment]
@mxnet-label-bot add [pr-awaiting-review]

**Comment by karan6181** [comment]
@apeforest Could you please review the PR? Thanks!

File: src/operator/tensor/elemwise_unary_op_trig.cc [hunk 1]
```diff
@@ -92,7 +92,7 @@ The storage type of ``cos`` output is always dense
 MXNET_OPERATOR_REGISTER_BINARY_WITH_SPARSE_CPU(_backward_cos, unary_bwd<mshadow_op::cos_grad>)
 .set_attr<nnvm::FGradient>("FGradient",
     [](const nnvm::NodePtr& n, const std::vector<nnvm::NodeEntry>& ograds) {
-      // ograds[0]: head_grad_grads (dL/dx_grad)
+      // ograds[0]: head_grad_grads (dL/dy_grad)
```
Comment (apeforest) [inline src/operator/tensor/elemwise_unary_op_trig.cc]: Why change this? I think the original dL/dx_grad is correct.

File: src/operator/tensor/elemwise_unary_op_trig.cc [hunk 2]
```diff
@@ -49,7 +49,7 @@ The storage type of ``sin`` output depends upon the input storage type:
 MXNET_OPERATOR_REGISTER_BINARY_WITH_SPARSE_CPU_DR(_backward_sin, unary_bwd<mshadow_op::sin_grad>)
 .set_attr<nnvm::FGradient>("FGradient",
     [](const nnvm::NodePtr& n, const std::vector<nnvm::NodeEntry>& ograds) {
-      // ograds[0]: head_grad_grads (dL/dxgrad)
+      // ograds[0]: head_grad_grads (dL/dy_grad)
```
Comment (apeforest) [inline src/operator/tensor/elemwise_unary_op_trig.cc]: Why change this? I think the original dL/dx_grad is correct.

File: src/operator/tensor/elemwise_unary_op_trig.cc [hunk 3]
```diff
@@ -257,7 +257,30 @@ The storage type of ``sinh`` output depends upon the input storage type:
 )code" ADD_FILELINE)
 .set_attr<nnvm::FGradient>("FGradient", ElemwiseGradUseIn{ "_backward_sinh" });
 
-MXNET_OPERATOR_REGISTER_BINARY_WITH_SPARSE_CPU_DR(_backward_sinh, unary_bwd<mshadow_op::sinh_grad>);
+MXNET_OPERATOR_REGISTER_BINARY_WITH_SPARSE_CPU_DR(_backward_sinh, unary_bwd<mshadow_op::sinh_grad>)
+.set_attr<nnvm::FGradient>("FGradient",
+    [](const nnvm::NodePtr& n, const std::vector<nnvm::NodeEntry>& ograds) {
+      // ograds[0]: dL/dy_grad
```
Comment (apeforest) [inline src/operator/tensor/elemwise_unary_op_trig.cc]: Why change this? I think the original dL/dx_grad is correct.

File: src/operator/tensor/elemwise_unary_op_trig.cc [hunk 4]
```diff
@@ -272,7 +295,31 @@ The storage type of ``cosh`` output is always dense
 )code" ADD_FILELINE)
 .set_attr<nnvm::FGradient>("FGradient", ElemwiseGradUseIn{ "_backward_cosh" });
 
-MXNET_OPERATOR_REGISTER_BINARY_WITH_SPARSE_CPU(_backward_cosh, unary_bwd<mshadow_op::cosh_grad>);
+MXNET_OPERATOR_REGISTER_BINARY_WITH_SPARSE_CPU(_backward_cosh, unary_bwd<mshadow_op::cosh_grad>)
+.set_attr<nnvm::FGradient>("FGradient",
+    [](const nnvm::NodePtr& n, const std::vector<nnvm::NodeEntry>& ograds) {
+      // ograds[0]: dL/dy_grad
```
Comment (apeforest) [inline src/operator/tensor/elemwise_unary_op_trig.cc]: Why change this? I think the original dL/dx_grad is correct.

File: src/operator/tensor/elemwise_unary_op_trig.cc (hunk 1 above)
Comment (kshitij12345) [reply src/operator/tensor/elemwise_unary_op_trig.cc]: Please have a look and let me know what you think.
https://github.com/apache/incubator-mxnet/pull/15331#discussion_r304541519

**Comment by karan6181** [comment]
@kshitij12345 is this PR good to go for merge?
@apeforest Could you please look into the updates? Thanks

**Comment by kshitij12345** [comment]
I guess we should wait till #15531 is in.

**Comment by kshitij12345** [comment]
@larroy @apeforest Gentle ping for review.

**Pull request merged**